src/pybind/mgr/rook/module.py

```diff
@@ -0,0 +1,392 @@
+            elif service_type == "mds":
+                # MDS daemon names are the tail of the pod name with
+                # an 'm' prefix.
+                # TODO: Would be nice to get this out a label though.
+                sl.daemon_name = "m" + sl.container_id.split("-")[-1]
+            elif service_type == "mon":
+                sl.daemon_name = p['labels']["mon"]
+            elif service_type == "mgr":
+                # FIXME: put a label on the pod to consume
+                # from here
+                raise NotImplementedError("mgr")
+            elif service_type == "rgw":
+                # FIXME: put a label on the pod to consume
+                # from here
+                raise NotImplementedError("rgw")
+
+            result.locations.append(sl)
+
+        return result
+
+    def add_stateless_service(self, service_type, spec):
+        # assert isinstance(spec, orchestrator.StatelessServiceSpec)
+
+        if service_type == "mds":
+            return RookWriteCompletion(
+                lambda: self.rook_cluster.add_filesystem(spec), None,
+                "Creating Filesystem services for {0}".format(spec.name))
+        else:
+            # TODO: RGW, NFS
+            raise NotImplementedError(service_type)
+
+    def create_osds(self, spec):
+        # Validate spec.node
+        if not self.rook_cluster.node_exists(spec.node):
+            raise RuntimeError("Node '{0}' is not in the Kubernetes "
+                               "cluster".format(spec.node))
+
+        # Validate whether cluster CRD can accept individual OSD
+        # creations (i.e. not useAllDevices)
+        if not self.rook_cluster.can_create_osd():
+            raise RuntimeError("Rook cluster configuration does not "
+                               "support OSD creation.")
+
+        def execute():
+            self.rook_cluster.add_osds(spec)
+
+        def is_complete():
+            # Find OSD pods on this host
+            pod_osd_ids = set()
+            pods = self._k8s.list_namespaced_pod("rook-ceph",
+                                                 label_selector="rook_cluster=rook-ceph,app=rook-ceph-osd",
+                                                 field_selector="spec.nodeName={0}".format(
+                                                     spec.node
+                                                 )).items
+            for p in pods:
+                pod_osd_ids.add(int(p.metadata.labels['ceph-osd-id']))
+
+            self.log.debug('pod_osd_ids={0}'.format(pod_osd_ids))
+
+            found = []
+            osdmap = self.get("osd_map")
+            for osd in osdmap['osds']:
+                osd_id = osd['osd']
+                if osd_id not in pod_osd_ids:
+                    continue
+
+                metadata = self.get_metadata('osd', "%s" % osd_id)
+                if metadata and metadata['devices'] in spec.drive_group.devices:
+                    found.append(osd_id)
```

Comment:
shall we just `break` or return `True` here?

Reply:
We could do that currently, but I'd rather leave this code in for later extending to handle multi-OSD drive groups nicely (currently this code just goes complete as soon as one OSD has been created, it will need improving later)